packages/compass-aggregations/src/modules/explain.ts

```diff
@@ -1,51 +1,217 @@
+  error: string;
+};
+
+type ExplainCancelledAction = {
+  type: ActionTypes.ExplainCancelled;
 };
 
 export type Actions =
-  | ModalOpenedAction
-  | ModalClosedAction;
+  | ExplainStartedAction
+  | ExplainFinishedAction
+  | ExplainFailedAction
+  | ExplainCancelledAction;
+
+export type ExplainData = {
+  plan: Document;
+  stats?: {
+    executionTimeMillis: number;
+    nReturned: number;
+    usedIndexes: IndexInformation[];
+  };
+};
 
 export type State = {
+  isLoading: boolean;
   isModalOpen: boolean;
+  explain?: ExplainData;
+  abortController?: AbortController;
+  error?: string;
 };
 
 export const INITIAL_STATE: State = {
+  isLoading: false,
   isModalOpen: false,
 };
 
 const reducer: Reducer<State, Actions> = (state = INITIAL_STATE, action) => {
   switch (action.type) {
-    case ActionTypes.ModalOpened:
+    case ActionTypes.ExplainStarted:
       return {
         isModalOpen: true,
+        isLoading: true,
+        explain: undefined,
+        error: undefined,
+        abortController: action.abortController,
       };
-    case ActionTypes.ModalClosed:
+    case ActionTypes.ExplainFinished:
       return {
-        isModalOpen: false,
+        isModalOpen: true,
+        isLoading: false,
+        explain: action.explain,
+        error: undefined,
+        abortController: undefined,
       };
+    case ActionTypes.ExplainFailed:
+      return {
+        isModalOpen: true,
+        isLoading: false,
+        explain: undefined,
+        error: action.error,
+        abortController: undefined,
+      };
+    case ActionTypes.ExplainCancelled:
+      return INITIAL_STATE;
     default:
       return state;
   }
 };
 
-export const openExplainModal = (): ModalOpenedAction => ({
-  type: ActionTypes.ModalOpened,
-});
+export const closeExplainModal = (): ThunkAction<
+  void,
+  RootState,
+  void,
+  Actions
+> => {
+  return (dispatch) => {
+    dispatch(cancelExplain());
+  };
+};
+
+export const cancelExplain = (): ThunkAction<
+  void,
+  RootState,
+  void,
+  Actions
+> => {
+  return (dispatch, getState) => {
+    const { explain: { abortController } } = getState();
+    abortController?.abort();
+    dispatch({
+      type: ActionTypes.ExplainCancelled
+    });
+  };
+};
+
+export const explainAggregation = (): ThunkAction<
+  void,
+  RootState,
+  void,
+  Actions
+> => {
+  return async (dispatch, getState) => {
+    const {
+      pipeline,
+      namespace,
+      maxTimeMS,
+      collation,
+      dataService: { dataService },
+    } = getState();
 
-export const closeExplainModal = (): ModalClosedAction => ({
-  type: ActionTypes.ModalClosed,
-});
+    if (!dataService) {
+      return;
+    }
+
+    try {
+      const abortController = new AbortController();
+      const signal = abortController.signal;
+      dispatch({
+        type: ActionTypes.ExplainStarted,
+        abortController,
+      });
+
+      const options: AggregateOptions = {
+        maxTimeMS: maxTimeMS || DEFAULT_MAX_TIME_MS,
```

Comment:
Zero is a valid `maxTimeMS` value

```suggestion
        maxTimeMS: maxTimeMS ?? DEFAULT_MAX_TIME_MS,
```